### DISCUSSION TITLE
Como acessar o Token em cada requisição?

### DISCUSSION THREAD

**Discussion by Hawkinks** [OpenSourceCommunityBrasil/REST-DataWare#359] [Q&A]
Meu plano é embutir as permissões do usuário no Token, para que toda vez que ele fizer uma requisição, o servidor verifique se ele possui as permissões necessárias.
Estou utilizando autenticação BearerToken.
Tentei pelo método OnAuthRequest do evento no ServerEvents, mas ele não passa o Token por ali.
Onde eu conseguiria acessar o Token? Ou há uma melhor abordagem para isso?

Versão: 2.1
Delphi

**Comment by mobius1qwe** [comment]
Na versão 2.1, no momento, o token tá "funcionando" mas vai sofrer uma reformulada pra ficar no padrão JWT da RFC 7519.
Pra já ir se preparando pro futuro, utilze cripto HMAC256, com o tipo rdwJWT nas propriedades do componente.

Daí, no ServerMethodDataModule, no evento OnUserTokenAuth você vai manipular os dados da autenticação e fazer as validações do seu jeito.

Pra incluir informações no token, basta acrescentar no payload essas informações. Enviando por POST que é o verbo correto de acordo com a RFC. Daí no evento OnUserTokenAuth você decodifica o base64 do token e pega o JSON com os dados.

Você também poderia enviar essas informações no cabeçalho da requisição (header) e pegar no mesmo evento através do `DWParams.ItemsString['nome do param'].AsString`.

**Reply by Hawkinks** [reply]
Entendi. Daí fazendo essa validação pelo OnUserTokenAuth, como faço para saber qual evento (do ServerEvents) o cliente está requisitando?

Porque pra validar as permissões eu preciso saber qual evento o cliente está querendo acessar.

**Reply by mobius1qwe** [reply]
No momento não tem esse valor nos parâmetros do método, mas é algo que pode ser incluído.

**Reply by Hawkinks** [reply]
Acha que devo abrir um feature request?

**Reply by mobius1qwe** [reply]
pode ser

**Reply by Hawkinks** [reply]
Se me permite perguntar aqui.
O Rest Dataware possui alguma forma nativa de gerenciar sessões?
